### DISCUSSION TITLE
with_polymorphic and with_for_update

### DISCUSSION THREAD

**Discussion by valq7711** [sqlalchemy/sqlalchemy#11952] [Usage Questions] [answered]
Hi! I use `joined table inheritance` and  need to lock records `for update` in **root table**
(SQLAlchemy 1.4.54)

#### with_for_update()
```python
poly = with_polymorphic(SomeBase, [Foo, Bar], flat=True)
q = select(poly).with_for_update()
```
it doesn't work: `sqlalchemy.exc.DBAPIError: (sqlalchemy.dialects.postgresql.asyncpg.Error)
<class 'asyncpg.exceptions.FeatureNotSupportedError'>: FOR UPDATE cannot be applied to the nullable side of an outer join
`

#### with_for_update(of=poly)
```python
q = select(poly).with_for_update(of=poly)
```
it also doesn't work because it compiled to `... FOR UPDATE foo_1, bar_1, some_base_1`   - explicit equivalent of just `FOR UPDATE`

#### with_for_update(of=`<RootTableClass>`)
```python
q = select(poly).with_for_update(of=SomeBase)
```
it also doesn't work because `sqlalchemy.exc.ProgrammingError: (sqlalchemy.dialects.postgresql.asyncpg.ProgrammingError) <class 'asyncpg.exceptions.UndefinedTableError'>: relation "some_base" in FOR UPDATE clause not found in FROM clause`

**So how do I reference the root table alias?** There are only `poly.Foo, poly.Bar` but not `poly.SomeBase`

At the moment I have to use a hard-coded workaround:
```python
q = select(poly).with_for_update(of='some_base_1')
```

Full example of the problem:
```python
from enum import Enum

from sqlalchemy import Column, String, ForeignKey, select
from sqlalchemy.dialects import postgresql
from sqlalchemy.orm import with_polymorphic, declarative_base


Base = declarative_base()


class FooBarType(str, Enum):
    FOO = 'FOO'
    BAR = 'BAR'


class SomeBase(Base):
    __tablename__ = 'some_base'

    id = Column(String(), primary_key=True)
    super_type = Column(postgresql.ENUM(FooBarType, name='foo_bar_type'))

    __mapper_args__ = {
        "polymorphic_on": super_type
    }


class Foo(SomeBase):
    __tablename__ = 'foo'

    id = Column(String(), ForeignKey(SomeBase.id), primary_key=True)

    __mapper_args__ = {
        "polymorphic_identity": FooBarType.FOO
    }


class Bar(SomeBase):
    __tablename__ = 'bar'

    id = Column(String(), ForeignKey(SomeBase.id), primary_key=True)

    __mapper_args__ = {
        "polymorphic_identity": FooBarType.BAR
    }


poly = with_polymorphic(SomeBase, '*', flat=True)
for update_of in [None, poly, SomeBase, SomeBase.__table__, poly.__table__]:
    q = select(poly).with_for_update(of=update_of)
    stmt = q.compile(dialect=postgresql.dialect(), compile_kwargs={"literal_binds": True})
    print(stmt, '\n')

```
output:
```
SELECT some_base_1.id, some_base_1.super_type, foo_1.id AS id_1, bar_1.id AS id_2 
FROM some_base AS some_base_1 LEFT OUTER JOIN foo AS foo_1 ON some_base_1.id = foo_1.id LEFT OUTER JOIN bar AS bar_1 ON some_base_1.id = bar_1.id 
FOR UPDATE 

SELECT some_base_1.id, some_base_1.super_type, foo_1.id AS id_1, bar_1.id AS id_2
FROM some_base AS some_base_1 LEFT OUTER JOIN foo AS foo_1 ON some_base_1.id = foo_1.id LEFT OUTER JOIN bar AS bar_1 ON some_base_1.id = bar_1.id
FOR UPDATE OF bar_1, foo_1, some_base_1

SELECT some_base_1.id, some_base_1.super_type, foo_1.id AS id_1, bar_1.id AS id_2
FROM some_base AS some_base_1 LEFT OUTER JOIN foo AS foo_1 ON some_base_1.id = foo_1.id LEFT OUTER JOIN bar AS bar_1 ON some_base_1.id = bar_1.id 
FOR UPDATE OF some_base

SELECT some_base_1.id, some_base_1.super_type, foo_1.id AS id_1, bar_1.id AS id_2
FROM some_base AS some_base_1 LEFT OUTER JOIN foo AS foo_1 ON some_base_1.id = foo_1.id LEFT OUTER JOIN bar AS bar_1 ON some_base_1.id = bar_1.id 
FOR UPDATE OF some_base

SELECT some_base_1.id, some_base_1.super_type, foo_1.id AS id_1, bar_1.id AS id_2
FROM some_base AS some_base_1 LEFT OUTER JOIN foo AS foo_1 ON some_base_1.id = foo_1.id LEFT OUTER JOIN bar AS bar_1 ON some_base_1.id = bar_1.id 
FOR UPDATE OF some_base
```

**Comment by zzzeek** [comment]
hi -

probably like this: `q = select(poly).with_for_update(of=SomeBase.__table__)`

but I'd need the models to try it

**Comment by valq7711** [comment]
> probably like this: `q = select(poly).with_for_update(of=SomeBase.__table__)`

@zzzeek thanks for your answer, but unfortunately this doesn't work either :( - `... FOR UPDATE OF some_base` - not alias, just table name
> but I'd need the models to try it

I added full example of the problem

Also I tried this:
```python
root_aliased = aliased(SomeBase)
poly = with_polymorphic(root_aliased, '*', flat=True, aliased=True)
q = select(poly).with_for_update(of=root_aliased)
```

and it is compiled to almost what I need, but look at `SELECT` and `FROM` clauses:
```
SELECT some_base_1.id, some_base_1.super_type
FROM some_base AS some_base_1, some_base AS some_base_2 LEFT OUTER JOIN foo AS foo_1 ON some_base_2.id = foo_1.id LEFT OUTER JOIN bar AS bar_1 ON some_base_2.id = bar_1.id FOR UPDATE OF some_base_1
```
and it seems like a bug, not related to `with_for_update` at all

**Comment by zzzeek** [comment, marked as answer]
so with_polymorphic can either be given the SELECT it should use, or it will generate one for you.  once it generates that, it doesnt provide simple accessors for the components that created that SELECT, since there's virtually (except for this exact case) never a need for it.      You can generate your own SELECT by setting up that whole join manually and keeping track of the original base table, but that's pretty tedious.

So when with_polymorphic generates the SELECT, there is a way to get at the individual tables and it's via the APIs that the ORM uses to line up columns; it's column-oriented so you have to convert back and forth, for this case it looks like this:

```py
from sqlalchemy import inspect
poly = with_polymorphic(SomeBase, '*', flat=True)
lead_table = inspect(poly).selectable.corresponding_column(SomeBase.id).table
q = select(poly).with_for_update(of=lead_table)
```

that then runs fine

```
2024-10-05 09:19:33,917 INFO sqlalchemy.engine.Engine SELECT some_base_1.id, some_base_1.super_type, foo_1.id AS id_1, bar_1.id AS id_2 
FROM some_base AS some_base_1 LEFT OUTER JOIN foo AS foo_1 ON some_base_1.id = foo_1.id LEFT OUTER JOIN bar AS bar_1 ON some_base_1.id = bar_1.id FOR UPDATE OF some_base_1

```

I can see helper functions to make this more succinct but I dont know that they'd be any easier in practice since they'd be just as obscure as `corresponding_column()`

**Reply by valq7711** [reply]
Thank you very much!
